app/controllers/UserController.java

```diff
@@ -68,16 +70,20 @@ public Result confirm(String registrationToken){
      * @return JSON which contains authToken is returned if successful.
      */
     public Result login(){
-         UserSession userSession = play.libs.Json.fromJson(request().body().asJson(), UserSession.class);
+        UserSession userSession = play.libs.Json.fromJson(request().body().asJson(), UserSession.class);
 
+        if(!userSession.isReadyForLogin()){
+            return badRequest(new Gson().toJson(new Error(Resources.BAD_REQUEST_INSUFFICIENT_DATA)));
+        }
         User user = manager.getUserFromSession(userSession);
 
-        if(user == null){
+        if(!user.exists()){
             return unauthorized(new Gson().toJson(new Error(Resources.UNAUTHORIZED_NO_EMAIL)));
         }
         if( !user.isValidLoginInfo()) {
             return unauthorized(new Gson().toJson(new Error(Resources.UNAUTHORIZED_INPUT_DOES_NOT_MATCH)));
         }
+        user = manager.matchTheUser(userSession);
 
         return ok(jsonOutput.render(new Gson().toJson(new TokenHelper(user.getAuthToken().getToken()))));
     }
```

Comment:
Think about the flow in this method, because it's no good at all! (multiple unnecessary db calls, wrong validations and error messages...)